### PR TITLE
feat: implement build in source for custom make

### PR DESCRIPTION
**Pull request by torresxb1**
*Issue #, if available:*

*Description of changes:*
Adds the ability to build in source for custom make. 

**Note**: This was branched off another PR: https://github.com/aws/aws-lambda-builders/pull/416. While the parent PR is not merged in, those changes will also show up in this PR's diff. See this PR's actual changes here: https://github.com/torresxb1/aws-lambda-builders/compare/in-source-entry...torresxb1:aws-lambda-builders:in-source-custom-make

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws_lambda_builders/workflows/custom_make/workflow.py [hunk 1]
```diff
@@ -45,17 +44,43 @@ def __init__(self, source_dir, artifacts_dir, scratch_dir, manifest_path, runtim
...
-        self.actions = [CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES), make_action]
+        self.actions = []
+
+        if not self.build_in_source:
+            # if we're building on scratch_dir, we have to first copy the source there
+            self.actions.append(CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES))
+
+        self.actions.append(make_action)
+
+    def _get_working_directory(self, options, source_dir, scratch_dir, build_in_source):
```
Comment (mildaniel) [inline aws_lambda_builders/workflows/custom_make/workflow.py:71]: nit: Can we add type hints for newer functions. We're trying to be better about it going forward in this repo.

**Review by mildaniel — APPROVED** [review]
Thanks Rup, looks good. Can we just get a small unit test for the new `_get_working_directory` function 🙏

**Comment by lucashuy** [comment]
Looks good to me, will defer to Daniel's comments

File: aws_lambda_builders/workflows/custom_make/workflow.py (hunk 1 above)
Comment (torresxb1) [reply aws_lambda_builders/workflows/custom_make/workflow.py:71]: for sure, will do.

**Comment by torresxb1** [comment]
> Thanks Rup, looks good. Can we just get a small unit test for the new _get_working_directory function 🙏

@mildaniel This is actually something I've gone back and forth on in the past and still unsure about: should we unit test private functions or only public? It seems like it's a topic that has people on both sides, eg.: https://stackoverflow.com/questions/105007/should-i-test-private-methods-or-only-public-ones

In this case, I tested `_get_working_directory` indirectly by adding tests for the workflow itself (`test_build_in_source` and `test_build_in_source_with_custom_working_directory`). I have tested private functions before, so I haven't been consistent with this either. Is there a consensus on where the team stands on this?

The main point I see for not testing private functions is that we'd be testing implementation details rather than behaviour, and also refactoring private functions could mean updating unit tests even if the public behaviour didn't change.

**Comment by mildaniel** [comment]
My opinion is that all functions should be unit tested, public or private. I see both arguments, but I think it's a lot easier to cover edge cases on a granular level for each function than it is to try to cover every possible case on a macro level. If we have a module where only 5% of the code is exposed publicly, it's very difficult to write tests to cover all possible scenarios indirectly for the other 95% of code. Coverage will decrease, and ultimately so will the quality of the tests.

This is just my opinion though, so I'm happy to hear what others think.

File: aws_lambda_builders/workflows/custom_make/workflow.py [hunk 2]
```diff
@@ -45,17 +44,43 @@ def __init__(self, source_dir, artifacts_dir, scratch_dir, manifest_path, runtim
 
         subprocess_make = SubProcessMake(make_exe=self.binaries["make"].binary_path, osutils=self.os_utils)
 
+        build_in_source = kwargs.get("build_in_source")
```
Comment (sriram-mv) [inline aws_lambda_builders/workflows/custom_make/workflow.py:47]: nit: can we move all the kwargs extraction together.

**Review by sriram-mv — CHANGES_REQUESTED** [review]

**Comment by torresxb1** [comment]
@mildaniel let me bring it up to the team to see if there's strong opinions. Won't block it on this, but I'll see if there's some consensus.

File: aws_lambda_builders/workflows/custom_make/workflow.py (hunk 2 above)
Comment (torresxb1) [reply aws_lambda_builders/workflows/custom_make/workflow.py]: done

File: aws_lambda_builders/workflows/custom_make/workflow.py [hunk 3]
```diff
@@ -45,17 +44,43 @@ def __init__(self, source_dir, artifacts_dir, scratch_dir, manifest_path, runtim
...
-        self.actions = [CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES), make_action]
+        self.actions = []
+
+        if not self.build_in_source:
+            # if we're building on scratch_dir, we have to first copy the source there
+            self.actions.append(CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES))
+
+        self.actions.append(make_action)
+
+    def _get_working_directory(self, options: dict, source_dir: str, scratch_dir: str, build_in_source: bool):
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply aws_lambda_builders/workflows/custom_make/workflow.py]: done

File: aws_lambda_builders/workflows/custom_make/workflow.py [hunk 4]
```diff
@@ -45,17 +44,43 @@ def __init__(self, source_dir, artifacts_dir, scratch_dir, manifest_path, runtim
...
-            scratch_dir,
             manifest_path,
             osutils=self.os_utils,
             subprocess_make=subprocess_make,
             build_logical_id=build_logical_id,
             working_directory=working_directory,
         )
 
-        self.actions = [CopySourceAction(source_dir, scratch_dir, excludes=self.EXCLUDED_FILES), make_action]
+        self.actions = []
```
[the remark this answers is not in the document]
Comment (torresxb1) [reply aws_lambda_builders/workflows/custom_make/workflow.py:65]: not sure how, given that one item is only added conditionally 🤔

[the remark this answers is not in the document]
Comment (sriram-mv) [reply aws_lambda_builders/workflows/custom_make/workflow.py:65]: self.actions = [CopySrc(...), make_action] if .... else [make_action]

[the remark this answers is not in the document]
Comment (sriram-mv) [reply aws_lambda_builders/workflows/custom_make/workflow.py:65]: Approving without blocking.

**Review by sriram-mv — APPROVED** [review]

File: aws_lambda_builders/workflows/custom_make/workflow.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (torresxb1) [reply aws_lambda_builders/workflows/custom_make/workflow.py:65]: Merging without this change as I'm not sure if it improves readability and I feel like it wouldn't be able to nicely handle adding more conditional actions in the future.
